src/common/collection.c

```diff
@@ -1830,7 +1913,24 @@ int dt_collection_image_offset(int imgid)
   return dt_collection_image_offset_with_collection(darktable.collection, imgid);
 }
 
-static void _dt_collection_recount_callback_1(gpointer instace, gpointer user_data)
+static gboolean _dt_collection_free_list(gpointer user_data)
+{
+  if(!user_data) return FALSE;
+  GList *l = user_data;
+  g_list_free(l);
+  return FALSE; // dont call again
+}
+static void _dt_collection_changed_callback(gpointer instance, dt_collection_change_t query_change, gpointer imgs,
```

Comment:
yes, I'm not fully satisfied of this either. The problem is that `DT_SIGNAL_COLLECTION_CHANGED` has as an argument a `GList` containing the ids of affected images. We then need to free this list at some point to avoid memory leak.
iiuc `DT_SIGNAL_DEVELOP_HISTORY_WILL_CHANGE` as the same list problem too. it's solved there by freeing the list in the unique connected function.
But here, we have multiple connected functions. For instance only one used the list, but we can't be sure that it'll still be the case in the future.

So I've seen only 2 solutions for freeing the list : 
1. freeing it in `_dt_collection_changed_callback` (`thumbtable.c:997`) keeping in mind that the list we be invalid in other connecting function.
2. freeing the list in a special connected function with a long enough timeout to be reasonably sure that it'll not be in use (the list is not big in memory, so imho, it's not a problem to keep it in memory for some time)

I've not take the 1 route, as imo there's good chances that we'll need to use this list in other functions in a very near future (culling/preview, maps, ...)
For 2., I think the risk of timeout crash is really low, as it would mean that we have a connected function that takes more than 15s. to execute. and even in this case, it can be avoided by copying the list at the beginning of the function...

Reply:
Or you should implement correct link counting, e.g. by wrapping GList into GObject. Freeing memory on timeout like that is not something anyone should ever even think about, much less put into actual code.

Reply:
@AlicVB : Maybe a third solution (which would also "fix" the fragile implementation of DT_SIGNAL_DEVELOP_HISTORY_WILL_CHANGE). Have a way to register a cleaning routine for each signal kind. This free routine would be called when all the signals callbacks have been called. Not sure how easy this is, but it seems a generic implementation to fix this.

How does that sound?

Reply:
@AlicVB : My solution won't work as we do not control the way the signals are called. So we probably need to count the number of signal connected dt_control_signal_connect() decrement in each callback the counter and free when 0 is reached. A stupid ref counter, probably what @parafin had in mind.

But we probably also want to take into account dt_control_signal_block_by_func. Which is non trivial I suspect...